python/pylibcudf/pylibcudf/strings/combine.pyx

```diff
@@ -0,0 +1,206 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.
+from libcpp.memory cimport unique_ptr
+from libcpp.utility cimport move
+from pylibcudf.column cimport Column
+from pylibcudf.libcudf.column.column cimport column
+from pylibcudf.libcudf.scalar.scalar cimport string_scalar
+from pylibcudf.libcudf.strings cimport combine as cpp_combine
+from pylibcudf.scalar cimport Scalar
+from pylibcudf.table cimport Table
+
+from cython.operator import dereference
+from pylibcudf.libcudf.strings.combine import \
+    output_if_empty_list as OutputIfEmptyList  # no-cython-lint
+from pylibcudf.libcudf.strings.combine import \
+    separator_on_nulls as SeparatorOnNulls  # no-cython-lint
+
+
+cpdef Column concatenate(
+    Table strings_columns,
+    ColumnOrScalar separator,
+    Scalar narep,
+    Scalar col_narep,
+    separator_on_nulls separate_nulls,
+):
+    """
+    Concatenates all strings in the column into one new string delimited
+    by an optional separator string.
+
+    Parameters
+    ----------
+    strings_columns : Table
+        Strings for this operation
+
+    separator : Column or Scalar
+        Separator(s) for a given row
+
+    narep : Scalar
+        String to replace a null separator for a given row.
+
+    col_narep : Scalar
+        String that should be used in place of any null strings found in any column.
+        Ignored when separator is a Scalar.
+
+    separate_nulls : SeparatorOnNulls
+        If YES, then the separator is included for null rows.
+
+    Returns
+    -------
+    Column
+        New column with concatenated results
+    """
+    cdef unique_ptr[column] c_result
+    cdef const string_scalar* c_narep = <const string_scalar*>(
+        narep.c_obj.get()
+    )
+    cdef const string_scalar* c_col_narep = <const string_scalar*>(
+        narep.c_obj.get()
+    )
+    cdef const string_scalar* c_separator
+
+    if ColumnOrScalar is Column:
+        with nogil:
+            c_result = move(
+                cpp_combine.concatenate(
+                    strings_columns.view(),
+                    separator.view(),
+                    dereference(c_narep),
```

Comment:
*issue*: there's currently no way of setting a pylibcudf's validity. So we can't access all of the libcudf functionality with this API if we already have a `Scalar`.

I think we need to be able to construct a `Scalar` whose `is_valid` method returns `False` to do that.

WDYT?

Reply:
Yes, that sounds right. In general we need much more full-featured constructors for Scalars (and Columns) than we currently have. I would support work to add that, but I would suggest that we open a separate issue to track so that this PR doesn't stall waiting for that feature.